// dbsc-toolkit/dist/core/testing/memory-storage-stub.js
export class MemoryStorage {
    sessions = new Map();
    keys = new Map();
    challenges = new Map();
    async getSession(id) {
        return this.sessions.get(id) ?? null;
    }
    async setSession(session) {
        this.sessions.set(session.id, session);
    }
    async deleteSession(id) {
        this.sessions.delete(id);
        this.keys.delete(`${id}:native`);
        this.keys.delete(`${id}:bound`);
    }
    async getBoundKey(sessionId, kind) {
        if (kind)
            return this.keys.get(`${sessionId}:${kind}`) ?? null;
        return (this.keys.get(`${sessionId}:native`) ??
            this.keys.get(`${sessionId}:bound`) ??
            null);
    }
    async setBoundKey(key) {
        this.keys.set(`${key.sessionId}:${key.kind}`, key);
    }
    async deleteBoundKey(sessionId, kind) {
        if (kind) {
            this.keys.delete(`${sessionId}:${kind}`);
            return;
        }
        this.keys.delete(`${sessionId}:native`);
        this.keys.delete(`${sessionId}:bound`);
    }
    async getChallenge(jti) {
        return this.challenges.get(jti) ?? null;
    }
    async setChallenge(challenge) {
        this.challenges.set(challenge.jti, challenge);
    }
    async consumeChallenge(jti) {
        const ch = this.challenges.get(jti);
        if (!ch || ch.consumed)
            return false;
        ch.consumed = true;
        return true;
    }
    async revokeSession(sessionId) {
        this.sessions.delete(sessionId);
        this.keys.delete(`${sessionId}:native`);
        this.keys.delete(`${sessionId}:bound`);
    }
    async revokeAllForUser(userId) {
        for (const [id, sess] of this.sessions.entries()) {
            if (sess.userId === userId) {
                this.sessions.delete(id);
                this.keys.delete(`${id}:native`);
                this.keys.delete(`${id}:bound`);
            }
        }
    }
}
