// dbsc-toolkit/dist/fastify/create-dbsc.js
import { randomBytes } from "node:crypto";
import { deriveSessionId, deviceCookieName, resolveCookieScope, } from "../core/index.js";
import { dbsc, bindSession } from "./index.js";
import { requireProof } from "./require-proof.js";
const DEVICE_COOKIE_TTL_SEC = 365 * 24 * 60 * 60;
/**
 * Returns a stable per-device value for the JWT `bind()` path: reads the
 * device cookie (named per cookie scope), or mints + sets one if absent.
 */
function resolveDeviceHint(reply, scope) {
    const name = deviceCookieName(scope);
    const { domain } = resolveCookieScope(scope);
    const existing = reply.request.cookies?.[name];
    if (existing)
        return existing;
    const value = randomBytes(16).toString("hex");
    reply.setCookie(name, value, {
        httpOnly: true,
        secure: scope.secure,
        sameSite: "lax",
        path: "/",
        maxAge: DEVICE_COOKIE_TTL_SEC,
        ...(domain !== undefined && { domain }),
    });
    return value;
}
/**
 * Builds a configured DBSC kit for Fastify. The static client SDK is not
 * mounted by `install()` — serve `dist/client/` yourself if you have a frontend.
 */
export function createDbsc(opts) {
    const secure = opts.secure ?? true;
    const cookieScope = opts.cookieScope;
    const cookieDomain = opts.cookieDomain;
    const scope = {
        secure,
        ...(cookieScope !== undefined && { cookieScope }),
        ...(cookieDomain !== undefined && { cookieDomain }),
    };
    const registrationPath = opts.registrationPath ?? "/dbsc/registration";
    async function bind(reply, a, b) {
        const bindOpts = typeof a === "string" ? b : a;
        let sessionId;
        if (typeof a === "string") {
            sessionId = a;
        }
        else {
            const deviceHint = bindOpts.deviceHint ?? resolveDeviceHint(reply, scope);
            sessionId = await deriveSessionId({
                userId: bindOpts.userId,
                deviceHint,
                ...(bindOpts.namespace !== undefined && { namespace: bindOpts.namespace }),
            });
        }
        await bindSession(reply, sessionId, opts.storage, {
            userId: bindOpts.userId,
            secure,
            ...(cookieScope !== undefined && { cookieScope }),
            ...(cookieDomain !== undefined && { cookieDomain }),
            registrationPath,
            ...(opts.registrationCookieTtl !== undefined && {
                registrationCookieTtl: opts.registrationCookieTtl,
            }),
            ...(opts.sessionTtl !== undefined && { sessionTtl: opts.sessionTtl }),
        });
        return sessionId;
    }
    return {
        async install(fastify) {
            if (!fastify.hasReplyDecorator("setCookie")) {
                const mod = (await import("@fastify/cookie"));
                await fastify.register((mod.default ?? mod));
            }
            await fastify.register(dbsc, opts);
            return fastify;
        },
        bind,
        requireProof,
    };
}
