// dbsc-toolkit/dist/fastify/index.js
import fp from "fastify-plugin";
import "@fastify/cookie";
import { handleRegistration, handleRefresh, handleBoundRegistration, handleBoundRefresh, issueChallenge, buildRegistrationHeader, buildChallengeHeader, readSessionResponseHeader, parseSessionSkippedHeader, REGISTRATION_HEADER, CHALLENGE_HEADER, LEGACY_REGISTRATION_HEADER, LEGACY_CHALLENGE_HEADER, NoopRateLimiter, emit, maybeEmitPolyfillMissing, resolveCookieNames, resolveCookieScope, cookieAttributesString, DbscProtocolError, DbscVerificationError, ErrorCodes, } from "../core/index.js";
export const DBSC_INTERNAL = Symbol("dbsc-toolkit.fastify.internal");
const cookieNames = (s) => resolveCookieNames(s);
const DEFAULT_BOUND_TTL_MS = 10 * 60 * 1000;
const DEFAULT_REG_TTL_MS = 24 * 60 * 60 * 1000;
const DEFAULT_SESSION_TTL_MS = 24 * 60 * 60 * 1000;
export async function bindSession(reply, sessionId, storage, opts) {
    const secure = opts.secure ?? true;
    const registrationPath = opts.registrationPath ?? "/dbsc/registration";
    const regCookieTtl = opts.registrationCookieTtl ?? DEFAULT_REG_TTL_MS;
    const sessionTtl = opts.sessionTtl ?? DEFAULT_SESSION_TTL_MS;
    const scope = {
        secure,
        ...(opts.cookieScope !== undefined && { cookieScope: opts.cookieScope }),
        ...(opts.cookieDomain !== undefined && { cookieDomain: opts.cookieDomain }),
    };
    const { domain } = resolveCookieScope(scope);
    const existing = await storage.getSession(sessionId);
    const now = Date.now();
    if (!existing) {
        await storage.setSession({
            id: sessionId,
            userId: opts.userId,
            tier: "none",
            createdAt: now,
            expiresAt: now + sessionTtl,
            lastRefreshAt: 0,
        });
    }
    const COOKIES = cookieNames(scope);
    const challenge = await issueChallenge(sessionId, storage);
    const regHeader = buildRegistrationHeader({
        registrationPath,
        challenge: challenge.jti,
        cookieName: COOKIES.bound,
    });
    reply.header(REGISTRATION_HEADER, regHeader);
    reply.header(LEGACY_REGISTRATION_HEADER, regHeader);
    const cookieBase = {
        httpOnly: true,
        secure,
        sameSite: "lax",
        path: "/",
        ...(domain !== undefined && { domain }),
    };
    reply.setCookie(COOKIES.reg, sessionId, { ...cookieBase, maxAge: regCookieTtl / 1000 });
    reply.setCookie(COOKIES.challenge, challenge.jti, { ...cookieBase, maxAge: 5 * 60 });
}
const dbscPlugin = async (fastify, opts) => {
    const { storage, registrationPath = "/dbsc/registration", refreshPath = "/dbsc/refresh", boundStatePath = "/dbsc-bound/state", boundChallengePath = "/dbsc-bound/challenge", boundRegistrationPath = "/dbsc-bound/registration", boundRefreshPath = "/dbsc-bound/refresh", bound = true, boundCookieTtl = DEFAULT_BOUND_TTL_MS, refreshGraceMs = 30_000, registrationCookieTtl = DEFAULT_REG_TTL_MS, rateLimiter = new NoopRateLimiter(), replayCache, onEvent, autoBind, secure = true, cookieScope, cookieDomain, } = opts;
    const scope = {
        secure,
        ...(cookieScope !== undefined && { cookieScope }),
        ...(cookieDomain !== undefined && { cookieDomain }),
    };
    const { domain } = resolveCookieScope(scope);
    const cookieAttrs = cookieAttributesString(scope);
    const cookieOpts = {
        httpOnly: true,
        secure,
        sameSite: "lax",
        path: "/",
        ...(domain !== undefined && { domain }),
    };
    const COOKIES = cookieNames(scope);
    const polyfillMissingEmitted = new Set();
    fastify.decorateRequest("dbsc", null);
    fastify.addHook("onRequest", async (req, reply) => {
        const sessionId = req.cookies?.[COOKIES.bound] ?? null;
        const skipped = parseSessionSkippedHeader(req.headers);
        req.dbsc = {
            sessionId,
            tier: "none",
            skipped,
            revoke: async () => {
                if (sessionId)
                    await storage.revokeSession(sessionId);
                reply.clearCookie(COOKIES.bound, cookieOpts);
            },
        };
        req[DBSC_INTERNAL] = {
            storage,
            secure,
            boundEnabled: bound,
            ...(replayCache !== undefined && { replayCache }),
        };
        if (sessionId) {
            const session = await storage.getSession(sessionId);
            if (session) {
                const staleAfter = session.lastRefreshAt + boundCookieTtl + refreshGraceMs;
                const refreshable = session.tier === "dbsc" || session.tier === "bound";
                if (refreshable && Date.now() > staleAfter) {
                    req.dbsc.tier = "none";
                }
                else {
                    req.dbsc.tier = session.tier;
                }
                if (onEvent) {
                    await maybeEmitPolyfillMissing({
                        storage,
                        session,
                        ip: req.ip ?? "unknown",
                        emitted: polyfillMissingEmitted,
                        onEvent,
                    });
                }
            }
        }
        else if (autoBind && !req.cookies?.[COOKIES.reg]) {
            const result = await autoBind(req);
            if (result) {
                await bindSession(reply, result.sessionId, storage, {
                    userId: result.userId,
                    secure,
                    ...(cookieScope !== undefined && { cookieScope }),
                    ...(cookieDomain !== undefined && { cookieDomain }),
                    registrationPath,
                    registrationCookieTtl,
                });
            }
        }
    });
    fastify.post(registrationPath, async (req, reply) => {
        const ip = req.ip;
        const sessionId = req.cookies?.[COOKIES.reg];
        const expectedJti = req.cookies?.[COOKIES.challenge];
        if (!sessionId || !expectedJti) {
            return reply.status(400).send({ error: "missing session or challenge cookie" });
        }
        const allowed = await rateLimiter.checkRegistration(ip);
        if (!allowed)
            return reply.status(429).send({ error: "rate limited" });
        try {
            await handleRegistration({
                sessionId,
                secSessionResponseHeader: readSessionResponseHeader(req.headers),
                expectedJti,
            }, storage);
            emit(onEvent, {
                type: "registration",
                sessionId,
                tier: "dbsc",
                timestamp: Date.now(),
                algorithm: "ES256",
                ip,
            });
            reply.setCookie(COOKIES.bound, sessionId, {
                ...cookieOpts,
                maxAge: boundCookieTtl / 1000,
            });
            reply.clearCookie(COOKIES.challenge, cookieOpts);
            const origin = `${req.protocol}://${req.hostname}`;
            return reply.status(200).send({
                session_identifier: sessionId,
                refresh_url: refreshPath,
                scope: {
                    origin,
                    include_site: true,
                    scope_specification: [],
                },
                credentials: [
                    {
                        type: "cookie",
                        name: COOKIES.bound,
                        attributes: cookieAttrs,
                    },
                ],
            });
        }
        catch (err) {
            await rateLimiter.recordFailure(ip, sessionId);
            if (err instanceof DbscVerificationError || err instanceof DbscProtocolError) {
                return reply.status(400).send({ error: err.message });
            }
            throw err;
        }
    });
    fastify.post(refreshPath, async (req, reply) => {
        const ip = req.ip;
        const sessionIdHeader = req.headers["sec-secure-session-id"];
        const sessionId = (Array.isArray(sessionIdHeader) ? sessionIdHeader[0] : sessionIdHeader) ??
            req.cookies?.[COOKIES.bound];
        if (!sessionId)
            return reply.status(403).send();
        const allowed = await rateLimiter.checkRefresh(ip, sessionId);
        if (!allowed)
            return reply.status(429).send({ error: "rate limited" });
        const responseHeader = readSessionResponseHeader(req.headers);
        if (!responseHeader) {
            const challenge = await issueChallenge(sessionId, storage);
            reply.header(CHALLENGE_HEADER, buildChallengeHeader(challenge.jti, sessionId));
            reply.header(LEGACY_CHALLENGE_HEADER, buildChallengeHeader(challenge.jti, sessionId));
            reply.setCookie(COOKIES.challenge, challenge.jti, { ...cookieOpts, maxAge: 5 * 60 });
            return reply.status(403).send();
        }
        const expectedJti = req.cookies?.[COOKIES.challenge];
        if (!expectedJti) {
            const challenge = await issueChallenge(sessionId, storage);
            reply.header(CHALLENGE_HEADER, buildChallengeHeader(challenge.jti, sessionId));
            reply.header(LEGACY_CHALLENGE_HEADER, buildChallengeHeader(challenge.jti, sessionId));
            reply.setCookie(COOKIES.challenge, challenge.jti, { ...cookieOpts, maxAge: 5 * 60 });
            return reply.status(403).send();
        }
        try {
            await handleRefresh({ sessionId, secSessionResponseHeader: responseHeader, expectedJti }, storage);
            emit(onEvent, {
                type: "refresh",
                sessionId,
                tier: "dbsc",
                timestamp: Date.now(),
                ip,
            });
            reply.setCookie(COOKIES.bound, sessionId, { ...cookieOpts, maxAge: boundCookieTtl / 1000 });
            reply.clearCookie(COOKIES.challenge, cookieOpts);
            const origin = `${req.protocol}://${req.hostname}`;
            return reply.status(200).send({
                session_identifier: sessionId,
                refresh_url: refreshPath,
                scope: {
                    origin,
                    include_site: true,
                    scope_specification: [],
                },
                credentials: [
                    {
                        type: "cookie",
                        name: COOKIES.bound,
                        attributes: cookieAttrs,
                    },
                ],
            });
        }
        catch (err) {
            await rateLimiter.recordFailure(ip, sessionId);
            const stolenCheck = await storage.getBoundKey(sessionId);
            if (stolenCheck && err instanceof DbscVerificationError && err.code === ErrorCodes.SIGNATURE_INVALID) {
                emit(onEvent, { type: "session_stolen", sessionId, tier: "dbsc", timestamp: Date.now(), ip });
            }
            if (err instanceof DbscVerificationError || err instanceof DbscProtocolError) {
                emit(onEvent, {
                    type: "verification_failure",
                    sessionId,
                    tier: "dbsc",
                    timestamp: Date.now(),
                    reason: err.code,
                    ip,
                });
                const challenge = await issueChallenge(sessionId, storage);
                reply.header(CHALLENGE_HEADER, buildChallengeHeader(challenge.jti, sessionId));
                reply.header(LEGACY_CHALLENGE_HEADER, buildChallengeHeader(challenge.jti, sessionId));
                reply.setCookie(COOKIES.challenge, challenge.jti, { ...cookieOpts, maxAge: 5 * 60 });
                return reply.status(403).send({ error: err.message });
            }
            throw err;
        }
    });
    const readBoundSessionId = (req) => req.cookies?.[COOKIES.bound] ?? req.cookies?.[COOKIES.reg];
    fastify.get(boundStatePath, async (req, reply) => {
        reply.header("X-Server-Time", String(Date.now()));
        // Polyfill disabled: tell the client SDK to stand down.
        if (!bound) {
            return reply.status(200).send({ phase: "unbound", sessionId: null });
        }
        const skipped = parseSessionSkippedHeader(req.headers);
        const nativeSkipped = skipped.length ? skipped.map((s) => s.reason) : undefined;
        const sessionId = readBoundSessionId(req);
        if (!sessionId) {
            return reply.status(200).send({ phase: "unbound", sessionId: null, ...(nativeSkipped && { nativeSkipped }) });
        }
        const session = await storage.getSession(sessionId);
        if (!session) {
            return reply.status(200).send({ phase: "unbound", sessionId: null, ...(nativeSkipped && { nativeSkipped }) });
        }
        const nativeKey = await storage.getBoundKey(sessionId, "native");
        const boundKey = await storage.getBoundKey(sessionId, "bound");
        if (!nativeKey && !boundKey) {
            const challenge = await issueChallenge(sessionId, storage);
            return reply.status(200).send({
                phase: "needs-registration",
                sessionId,
                challenge: challenge.jti,
                ...(nativeSkipped && { nativeSkipped }),
            });
        }
        if (nativeKey && !boundKey) {
            const challenge = await issueChallenge(sessionId, storage);
            return reply.status(200).send({
                phase: "needs-bound-registration",
                sessionId,
                tier: session.tier,
                challenge: challenge.jti,
                refreshIntervalMs: boundCookieTtl,
                ...(nativeSkipped && { nativeSkipped }),
            });
        }
        return reply.status(200).send({
            phase: "bound",
            sessionId,
            tier: session.tier,
            refreshIntervalMs: boundCookieTtl,
            ...(nativeSkipped && { nativeSkipped }),
        });
    });
    if (bound) {
        fastify.get(boundChallengePath, async (req, reply) => {
            reply.header("X-Server-Time", String(Date.now()));
            const sessionId = readBoundSessionId(req);
            if (!sessionId)
                return reply.status(403).send({ error: "no session" });
            const session = await storage.getSession(sessionId);
            if (!session)
                return reply.status(403).send({ error: "no session" });
            const challenge = await issueChallenge(sessionId, storage);
            return reply.status(200).send({ challenge: challenge.jti });
        });
        fastify.post(boundRegistrationPath, async (req, reply) => {
            reply.header("X-Server-Time", String(Date.now()));
            const ip = req.ip;
            const allowed = await rateLimiter.checkRegistration(ip);
            if (!allowed)
                return reply.status(429).send({ error: "rate limited" });
            const sessionId = readBoundSessionId(req);
            if (!sessionId)
                return reply.status(400).send({ error: "missing session cookie" });
            const body = (req.body ?? {});
            if (!body.publicKey || !body.signature || !body.challenge) {
                return reply.status(400).send({ error: "publicKey, signature, and challenge are required" });
            }
            try {
                await handleBoundRegistration({ sessionId, publicKey: body.publicKey, signature: body.signature, expectedJti: body.challenge }, storage);
                emit(onEvent, {
                    type: "registration",
                    sessionId,
                    tier: "bound",
                    timestamp: Date.now(),
                    algorithm: "ES256",
                    ip,
                });
                reply.setCookie(COOKIES.bound, sessionId, { ...cookieOpts, maxAge: boundCookieTtl / 1000 });
                return reply.status(200).send({
                    session_identifier: sessionId,
                    refresh_url: boundRefreshPath,
                    tier: "bound",
                });
            }
            catch (err) {
                await rateLimiter.recordFailure(ip, sessionId);
                if (err instanceof DbscVerificationError || err instanceof DbscProtocolError) {
                    emit(onEvent, {
                        type: "verification_failure",
                        sessionId,
                        tier: "bound",
                        timestamp: Date.now(),
                        reason: err.code,
                        ip,
                    });
                    return reply.status(400).send({ error: err.message });
                }
                throw err;
            }
        });
        fastify.post(boundRefreshPath, async (req, reply) => {
            reply.header("X-Server-Time", String(Date.now()));
            const ip = req.ip;
            const sessionId = readBoundSessionId(req);
            if (!sessionId)
                return reply.status(403).send({ error: "no session" });
            const allowed = await rateLimiter.checkRefresh(ip, sessionId);
            if (!allowed)
                return reply.status(429).send({ error: "rate limited" });
            const body = (req.body ?? {});
            if (!body.challenge || !body.signature || typeof body.timestamp !== "number") {
                return reply.status(400).send({ error: "challenge, signature, and timestamp are required" });
            }
            try {
                await handleBoundRefresh({ sessionId, signature: body.signature, expectedJti: body.challenge, timestamp: body.timestamp }, storage);
                emit(onEvent, { type: "refresh", sessionId, tier: "bound", timestamp: Date.now(), ip });
                reply.setCookie(COOKIES.bound, sessionId, { ...cookieOpts, maxAge: boundCookieTtl / 1000 });
                return reply.status(200).send({
                    session_identifier: sessionId,
                    refresh_url: boundRefreshPath,
                    tier: "bound",
                });
            }
            catch (err) {
                await rateLimiter.recordFailure(ip, sessionId);
                const keyStillThere = await storage.getBoundKey(sessionId);
                if (keyStillThere && err instanceof DbscVerificationError && err.code === ErrorCodes.SIGNATURE_INVALID) {
                    emit(onEvent, { type: "session_stolen", sessionId, tier: "bound", timestamp: Date.now(), ip });
                }
                if (err instanceof DbscVerificationError || err instanceof DbscProtocolError) {
                    emit(onEvent, {
                        type: "verification_failure",
                        sessionId,
                        tier: "bound",
                        timestamp: Date.now(),
                        reason: err.code,
                        ip,
                    });
                    return reply.status(401).send({ error: err.message });
                }
                throw err;
            }
        });
    }
};
export const dbsc = fp(dbscPlugin, { name: "@dbsc-toolkit/server-fastify" });
export { requireBoundProof } from "./proof.js";
export { requireProof } from "./require-proof.js";
export { createDbsc } from "./create-dbsc.js";
