// dbsc-toolkit/dist/fastify/proof.js
import { verifyBoundProof, DbscVerificationError, } from "../core/index.js";
/** Gates a route on a fresh bound-key proof. */
export function requireBoundProof(opts) {
    const allowDbsc = opts.allowDbscWithoutProof ?? false;
    const signBody = opts.signBody ?? false;
    return async function preHandler(req, reply) {
        const dbsc = req.dbsc;
        if (!dbsc?.sessionId || dbsc.tier === "none") {
            reply.status(403).send({ error: "no active binding", tier: dbsc?.tier ?? "none" });
            return;
        }
        if (dbsc.tier === "dbsc" && allowDbsc)
            return;
        try {
            const proofHeaderRaw = req.headers["x-dbsc-bound-proof"];
            const proofHeader = Array.isArray(proofHeaderRaw) ? proofHeaderRaw[0] : proofHeaderRaw;
            let bodyBytes;
            if (signBody) {
                const raw = req.body;
                if (raw instanceof Buffer) {
                    bodyBytes = new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);
                }
                else if (raw instanceof Uint8Array) {
                    bodyBytes = raw;
                }
                else if (typeof raw === "string") {
                    bodyBytes = new TextEncoder().encode(raw);
                }
                else {
                    bodyBytes = new Uint8Array(0);
                }
            }
            await verifyBoundProof({
                sessionId: dbsc.sessionId,
                proofHeader,
                method: req.method,
                path: req.url.split("?")[0] ?? req.url,
                timestampWindowMs: opts.timestampWindowMs,
                signBody,
                bodyBytes,
                ...(opts.replayCache !== undefined && { replayCache: opts.replayCache }),
            }, opts.storage);
        }
        catch (err) {
            if (err instanceof DbscVerificationError) {
                reply.status(403).send({ error: err.message, code: err.code });
                return;
            }
            throw err;
        }
    };
}
