// dbsc-toolkit/dist/hono/create-dbsc.js
import { getCookie, setCookie } from "hono/cookie";
import { randomBytes } from "node:crypto";
import { deriveSessionId, deviceCookieName, resolveCookieScope, } from "../core/index.js";
import { dbsc, bindSession } from "./index.js";
import { requireProof } from "./require-proof.js";
const DEVICE_COOKIE_TTL_SEC = 365 * 24 * 60 * 60;
/**
 * Returns a stable per-device value for the JWT `bind()` path: reads the
 * device cookie (named per cookie scope), or mints + sets one if absent.
 */
function resolveDeviceHint(c, scope) {
    const name = deviceCookieName(scope);
    const { domain } = resolveCookieScope(scope);
    const existing = getCookie(c, name);
    if (existing)
        return existing;
    const value = randomBytes(16).toString("hex");
    setCookie(c, name, value, {
        httpOnly: true,
        secure: scope.secure,
        sameSite: "Lax",
        path: "/",
        maxAge: DEVICE_COOKIE_TTL_SEC,
        ...(domain !== undefined && { domain }),
    });
    return value;
}
/**
 * Builds a configured DBSC kit for Hono. The static client SDK is not mounted
 * by `install()` — serve `dist/client/` with your runtime's static handler.
 */
export function createDbsc(opts) {
    const secure = opts.secure ?? true;
    const cookieScope = opts.cookieScope;
    const cookieDomain = opts.cookieDomain;
    const scope = {
        secure,
        ...(cookieScope !== undefined && { cookieScope }),
        ...(cookieDomain !== undefined && { cookieDomain }),
    };
    const registrationPath = opts.registrationPath ?? "/dbsc/registration";
    const middleware = dbsc(opts);
    async function bind(c, a, b) {
        const bindOpts = typeof a === "string" ? b : a;
        let sessionId;
        if (typeof a === "string") {
            sessionId = a;
        }
        else {
            const deviceHint = bindOpts.deviceHint ?? resolveDeviceHint(c, scope);
            sessionId = await deriveSessionId({
                userId: bindOpts.userId,
                deviceHint,
                ...(bindOpts.namespace !== undefined && { namespace: bindOpts.namespace }),
            });
        }
        await bindSession(c, sessionId, opts.storage, {
            userId: bindOpts.userId,
            secure,
            ...(cookieScope !== undefined && { cookieScope }),
            ...(cookieDomain !== undefined && { cookieDomain }),
            registrationPath,
            ...(opts.registrationCookieTtl !== undefined && {
                registrationCookieTtl: opts.registrationCookieTtl,
            }),
            ...(opts.sessionTtl !== undefined && { sessionTtl: opts.sessionTtl }),
        });
        return sessionId;
    }
    return {
        middleware: () => middleware,
        install(app) {
            app.use(middleware);
            return app;
        },
        bind,
        requireProof,
    };
}
