// dbsc-toolkit/dist/express/create-dbsc.js
import express from "express";
import { fileURLToPath } from "node:url";
import { randomBytes } from "node:crypto";
import { deriveSessionId, parseCookieHeader, deviceCookieName, resolveCookieScope, } from "../core/index.js";
import { dbsc, bindSession } from "./index.js";
import { requireProof } from "./require-proof.js";
const DEVICE_COOKIE_TTL_MS = 365 * 24 * 60 * 60 * 1000;
/**
 * Returns a stable per-device value for the JWT `bind()` path: reads the
 * device cookie (named per cookie scope), or mints + sets one if absent. Two
 * browsers of the same user thus derive distinct sessionIds and bind
 * independently.
 */
function resolveDeviceHint(res, scope) {
    const name = deviceCookieName(scope);
    const { domain } = resolveCookieScope(scope);
    const req = res.req;
    const existing = req?.cookies?.[name] ??
        parseCookieHeader(req?.headers?.cookie)[name];
    if (existing)
        return existing;
    const value = randomBytes(16).toString("hex");
    res.cookie(name, value, {
        httpOnly: true,
        secure: scope.secure,
        sameSite: "lax",
        path: "/",
        maxAge: DEVICE_COOKIE_TTL_MS,
        ...(domain !== undefined && { domain }),
    });
    return value;
}
/**
 * Builds a configured DBSC kit. Storage, `secure`, TTLs, rate limiter and
 * telemetry are set once here; `install()`, `bind()` and `requireProof()` all
 * read this config — nothing is re-passed.
 */
export function createDbsc(opts) {
    const secure = opts.secure ?? true;
    const cookieScope = opts.cookieScope;
    const cookieDomain = opts.cookieDomain;
    const scope = {
        secure,
        ...(cookieScope !== undefined && { cookieScope }),
        ...(cookieDomain !== undefined && { cookieDomain }),
    };
    const clientPath = opts.clientPath ?? "/dbsc-client";
    const registrationPath = opts.registrationPath ?? "/dbsc/registration";
    const boundRegistrationPath = opts.boundRegistrationPath ?? "/dbsc-bound/registration";
    const boundRefreshPath = opts.boundRefreshPath ?? "/dbsc-bound/refresh";
    const middleware = dbsc(opts);
    async function bind(res, a, b) {
        const bindOpts = typeof a === "string" ? b : a;
        let sessionId;
        if (typeof a === "string") {
            sessionId = a;
        }
        else {
            // JWT path: no sessionId. An explicit deviceHint wins; otherwise the kit
            // manages a per-device cookie so each browser binds independently.
            const deviceHint = bindOpts.deviceHint ?? resolveDeviceHint(res, scope);
            sessionId = await deriveSessionId({
                userId: bindOpts.userId,
                deviceHint,
                ...(bindOpts.namespace !== undefined && { namespace: bindOpts.namespace }),
            });
        }
        await bindSession(res, sessionId, opts.storage, {
            userId: bindOpts.userId,
            secure,
            ...(cookieScope !== undefined && { cookieScope }),
            ...(cookieDomain !== undefined && { cookieDomain }),
            registrationPath,
            ...(opts.registrationCookieTtl !== undefined && {
                registrationCookieTtl: opts.registrationCookieTtl,
            }),
            ...(opts.sessionTtl !== undefined && { sessionTtl: opts.sessionTtl }),
        });
        return sessionId;
    }
    return {
        middleware: () => middleware,
        install(app) {
            if (opts.trustProxy !== false)
                app.set("trust proxy", true);
            const json = express.json();
            app.use(boundRegistrationPath, json);
            app.use(boundRefreshPath, json);
            app.use(middleware);
            if (clientPath !== false) {
                const clientDir = fileURLToPath(new URL("../client/", import.meta.url));
                app.use(clientPath, express.static(clientDir));
            }
            return app;
        },
        bind,
        requireProof,
    };
}
