// dbsc-toolkit/dist/express/proof.js
import { verifyBoundProof, DbscVerificationError, } from "../core/index.js";
/** Gates a route on a fresh bound-key proof. */
export function requireBoundProof(opts) {
    const allowDbsc = opts.allowDbscWithoutProof ?? false;
    const signBody = opts.signBody ?? false;
    return async (req, res, next) => {
        const dbsc = res.locals.dbsc;
        if (!dbsc?.sessionId || dbsc.tier === "none") {
            res.status(403).json({ error: "no active binding", tier: dbsc?.tier ?? "none" });
            return;
        }
        if (dbsc.tier === "dbsc" && allowDbsc) {
            next();
            return;
        }
        try {
            let bodyBytes;
            if (signBody) {
                const raw = req.body;
                if (raw instanceof Buffer) {
                    bodyBytes = new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);
                }
                else if (raw instanceof Uint8Array) {
                    bodyBytes = raw;
                }
                else if (typeof raw === "string") {
                    bodyBytes = new TextEncoder().encode(raw);
                }
                else {
                    bodyBytes = new Uint8Array(0);
                }
            }
            await verifyBoundProof({
                sessionId: dbsc.sessionId,
                proofHeader: req.headers["x-dbsc-bound-proof"],
                method: req.method,
                path: req.path,
                timestampWindowMs: opts.timestampWindowMs,
                signBody,
                bodyBytes,
                ...(opts.replayCache !== undefined && { replayCache: opts.replayCache }),
            }, opts.storage);
            next();
        }
        catch (err) {
            if (err instanceof DbscVerificationError) {
                res.status(403).json({ error: err.message, code: err.code });
                return;
            }
            throw err;
        }
    };
}
