// dbsc-toolkit/dist/client/installFetchInterceptor.js
import { wrapFetch } from "./wrapFetch.js";
/**
 * Replace `globalThis.fetch` (or the provided `fetch`) with a wrapper that
 * routes matching same-origin requests through `wrapFetch`, and everything
 * else through the original fetch. Returns an `uninstall` function that
 * restores the original.
 *
 * **When to use this:** apps with many guarded routes where calling
 * `wrapFetch` at every call site is a footgun. For small apps, the per-call
 * `wrapFetch(...)` shape stays the recommended default.
 *
 * **Safety:** the interceptor only touches requests whose URL is same-origin
 * and whose pathname matches one of `pathPrefixes`. Third-party hosts are
 * always passed through to the original fetch — the session key never leaves
 * your origin.
 */
export function installFetchInterceptor(opts) {
    if (!opts.pathPrefixes || opts.pathPrefixes.length === 0) {
        throw new Error("installFetchInterceptor: pathPrefixes must be a non-empty array");
    }
    for (const prefix of opts.pathPrefixes) {
        if (typeof prefix !== "string" || prefix.length === 0) {
            throw new Error("installFetchInterceptor: each prefix must be a non-empty string");
        }
        if (prefix === "/") {
            throw new Error("installFetchInterceptor: pathPrefixes cannot include '/' — specify explicit route prefixes (e.g. '/api/secure/') to avoid signing static assets, health checks, or third-party requests");
        }
        if (prefix.startsWith("http://") || prefix.startsWith("https://")) {
            throw new Error("installFetchInterceptor: pathPrefixes must be path-only (e.g. '/api/secure/') — absolute URLs are not allowed; the interceptor never signs cross-origin requests");
        }
        if (!prefix.startsWith("/")) {
            throw new Error("installFetchInterceptor: each prefix must start with '/' (e.g. '/api/secure/')");
        }
    }
    // Two distinct concepts: the upstream fetch the interceptor delegates to
    // (test-friendly override via opts.fetch), and the fetch that was on
    // globalThis at install time (what we restore on uninstall).
    const upstreamFetch = opts.fetch ?? globalThis.fetch.bind(globalThis);
    const priorGlobalFetch = globalThis.fetch;
    const signed = wrapFetch({
        fetch: upstreamFetch,
        ...(opts.signBody !== undefined && { signBody: opts.signBody }),
        ...(opts.headerName !== undefined && { headerName: opts.headerName }),
    });
    const interceptor = async (input, init) => {
        const url = resolveUrl(input);
        if (url && isSameOrigin(url) && matchesAnyPrefix(url.pathname, opts.pathPrefixes)) {
            return signed(input, init);
        }
        return upstreamFetch(input, init);
    };
    globalThis.fetch = interceptor;
    return () => {
        globalThis.fetch = priorGlobalFetch;
    };
}
function resolveUrl(input) {
    try {
        const raw = typeof input === "string" || input instanceof URL ? input.toString() : input.url;
        const base = typeof window !== "undefined" ? window.location.href : "http://localhost";
        return new URL(raw, base);
    }
    catch {
        return null;
    }
}
function isSameOrigin(url) {
    if (typeof window === "undefined")
        return true; // no DOM origin to compare against — assume same
    return url.origin === window.location.origin;
}
function matchesAnyPrefix(pathname, prefixes) {
    for (const p of prefixes) {
        if (pathname.startsWith(p))
            return true;
    }
    return false;
}
