// dbsc-toolkit/dist/nestjs/index.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var DbscModule_1;
import { Injectable, Module, Inject, ForbiddenException, InternalServerErrorException, } from "@nestjs/common";
import { verifyBoundProof, noBindingReason, DbscVerificationError, } from "../core/index.js";
import { dbsc, createDbsc, DBSC_INTERNAL, } from "../express/index.js";
export { bindSession } from "../express/index.js";
/** DI token for the DBSC options passed to `DbscModule.forRoot()`. */
export const DBSC_OPTIONS = Symbol("dbsc-toolkit.nestjs.options");
/** Injectable wrapper around the Express kit's `bind()` — call it in your login controller. */
let DbscService = class DbscService {
    kit;
    constructor(opts) {
        this.kit = createDbsc(opts);
    }
    /** Start a binding for an explicit session id. */
    bind(res, sessionId, opts) {
        return this.kit.bind(res, sessionId, opts);
    }
};
DbscService = __decorate([
    Injectable(),
    __param(0, Inject(DBSC_OPTIONS)),
    __metadata("design:paramtypes", [Object])
], DbscService);
export { DbscService };
/**
 * Registers DBSC on a NestJS app (Express platform). It mounts the protocol
 * middleware for every route and exposes `DbscService`:
 *
 *   @Module({ imports: [DbscModule.forRoot({ storage })] })
 *   export class AppModule {}
 */
let DbscModule = DbscModule_1 = class DbscModule {
    opts;
    constructor(opts) {
        this.opts = opts;
    }
    static forRoot(opts) {
        return {
            module: DbscModule_1,
            global: true,
            providers: [{ provide: DBSC_OPTIONS, useValue: opts }, DbscService],
            exports: [DBSC_OPTIONS, DbscService],
        };
    }
    configure(consumer) {
        consumer.apply(dbsc(this.opts)).forRoutes("*");
    }
};
DbscModule = DbscModule_1 = __decorate([
    Module({}),
    __param(0, Inject(DBSC_OPTIONS)),
    __metadata("design:paramtypes", [Object])
], DbscModule);
export { DbscModule };
async function runProof(req, res, opts) {
    const dbscLocals = res.locals.dbsc;
    const tier = dbscLocals?.tier ?? "none";
    const skipped = dbscLocals?.skipped ?? [];
    if (!dbscLocals?.sessionId || tier === "none") {
        throw new ForbiddenException({ error: "device-bound session required", currentTier: "none", reason: noBindingReason(skipped), skipped });
    }
    const internal = res.locals[DBSC_INTERNAL];
    const storage = opts.storage ?? internal?.storage;
    if (!storage) {
        throw new InternalServerErrorException("requireProof: storage unavailable — import DbscModule.forRoot(...)");
    }
    const allowDbscWithoutProof = opts.allowDbscWithoutProof ?? (internal?.boundEnabled === false ? true : undefined);
    if (tier === "dbsc" && allowDbscWithoutProof)
        return;
    let bodyBytes;
    const raw = req.body;
    if (raw instanceof Buffer)
        bodyBytes = new Uint8Array(raw.buffer, raw.byteOffset, raw.byteLength);
    else if (raw instanceof Uint8Array)
        bodyBytes = raw;
    else if (typeof raw === "string")
        bodyBytes = new TextEncoder().encode(raw);
    else
        bodyBytes = new Uint8Array(0);
    try {
        await verifyBoundProof({
            sessionId: dbscLocals.sessionId,
            proofHeader: req.headers["x-dbsc-bound-proof"],
            method: req.method,
            path: req.path,
            signBody: true,
            bodyBytes,
            ...(opts.timestampWindowMs !== undefined && { timestampWindowMs: opts.timestampWindowMs }),
            ...(internal?.replayCache !== undefined && { replayCache: internal.replayCache }),
        }, storage);
    }
    catch (err) {
        if (err instanceof DbscVerificationError) {
            throw new ForbiddenException({ error: err.message, code: err.code });
        }
        throw err;
    }
}
/**
 * Zero-config route guard. Requires a bound device + a per-request proof:
 *
 *   @UseGuards(DbscGuard)
 *   @Post("payment")
 *   pay() { ... }
 *
 * POST bodies are body-hashed, so a guarded POST must deliver raw bytes (set
 * `rawBody: true` in `NestFactory.create` and disable the JSON parser on that
 * route, or use a raw-body middleware). GET routes need no parser.
 */
let DbscGuard = class DbscGuard {
    async canActivate(context) {
        const http = context.switchToHttp();
        await runProof(http.getRequest(), http.getResponse(), {});
        return true;
    }
};
DbscGuard = __decorate([
    Injectable()
], DbscGuard);
export { DbscGuard };
/** Builds a guard with options baked in (e.g. a storage override or timestamp window). */
export function createDbscGuard(opts) {
    let ConfiguredDbscGuard = class ConfiguredDbscGuard {
        async canActivate(context) {
            const http = context.switchToHttp();
            await runProof(http.getRequest(), http.getResponse(), opts);
            return true;
        }
    };
    ConfiguredDbscGuard = __decorate([
        Injectable()
    ], ConfiguredDbscGuard);
    return ConfiguredDbscGuard;
}
