// dbsc-toolkit/dist/nextjs/create-dbsc.js
import { randomBytes } from "node:crypto";
import { deriveSessionId, deviceCookieName, resolveCookieScope, } from "../core/index.js";
const DEVICE_COOKIE_TTL_SEC = 365 * 24 * 60 * 60;
import { createDbscMiddleware, bindSession, getDbscSession, } from "./index.js";
import { requireProof } from "./require-proof.js";
/**
 * Builds a configured DBSC kit for Next.js. There is no `install()` — Next has
 * no app object; export `middleware()` from `middleware.ts` and call the rest
 * inside route handlers.
 */
export function createDbsc(opts) {
    const secure = opts.secure ?? true;
    const cookieScope = opts.cookieScope;
    const cookieDomain = opts.cookieDomain;
    const scope = {
        secure,
        ...(cookieScope !== undefined && { cookieScope }),
        ...(cookieDomain !== undefined && { cookieDomain }),
    };
    const registrationPath = opts.registrationPath ?? "/dbsc/registration";
    const mw = createDbscMiddleware(opts);
    function resolveDeviceHint(req, res) {
        const name = deviceCookieName(scope);
        const { domain } = resolveCookieScope(scope);
        const existing = req.cookies.get(name)?.value;
        if (existing)
            return existing;
        const value = randomBytes(16).toString("hex");
        res.cookies.set(name, value, {
            httpOnly: true,
            secure,
            sameSite: "lax",
            path: "/",
            maxAge: DEVICE_COOKIE_TTL_SEC,
            ...(domain !== undefined && { domain }),
        });
        return value;
    }
    async function bind(res, a, b) {
        const bindOpts = typeof a === "string" ? b : a;
        let sessionId;
        if (typeof a === "string") {
            sessionId = a;
        }
        else {
            const deviceHint = bindOpts.deviceHint ??
                (bindOpts.req ? resolveDeviceHint(bindOpts.req, res) : undefined);
            sessionId = await deriveSessionId({
                userId: bindOpts.userId,
                ...(deviceHint !== undefined && { deviceHint }),
                ...(bindOpts.namespace !== undefined && { namespace: bindOpts.namespace }),
            });
        }
        await bindSession(res, sessionId, opts.storage, {
            userId: bindOpts.userId,
            secure,
            ...(cookieScope !== undefined && { cookieScope }),
            ...(cookieDomain !== undefined && { cookieDomain }),
            registrationPath,
            ...(opts.registrationCookieTtl !== undefined && {
                registrationCookieTtl: opts.registrationCookieTtl,
            }),
            ...(opts.sessionTtl !== undefined && { sessionTtl: opts.sessionTtl }),
        });
        return sessionId;
    }
    return {
        middleware: () => mw,
        bind,
        getSession: (req, res) => getDbscSession(req, opts.storage, {
            ...(opts.boundCookieTtl !== undefined && { boundCookieTtl: opts.boundCookieTtl }),
            ...(opts.refreshGraceMs !== undefined && { refreshGraceMs: opts.refreshGraceMs }),
            ...(res !== undefined && { res }),
            ...(opts.onEvent !== undefined && { onEvent: opts.onEvent }),
            secure,
            ...(cookieScope !== undefined && { cookieScope }),
            ...(cookieDomain !== undefined && { cookieDomain }),
        }),
        requireProof: (req, session) => requireProof(req, session, {
            storage: opts.storage,
            ...(opts.bound !== undefined && { bound: opts.bound }),
            ...(opts.replayCache !== undefined && { replayCache: opts.replayCache }),
        }),
    };
}
