// dbsc-toolkit/dist/nextjs/proof.js
import { NextResponse } from "next/server.js";
import { verifyBoundProof, DbscVerificationError, } from "../core/index.js";
/** Gates a route on a fresh bound-key proof. Use inside an App Router handler. */
export async function requireBoundProof(req, session, opts) {
    const allowDbsc = opts.allowDbscWithoutProof ?? false;
    if (!session.sessionId || session.tier === "none") {
        return {
            ok: false,
            response: NextResponse.json({ error: "no active binding", tier: session.tier }, { status: 403 }),
        };
    }
    if (session.tier === "dbsc" && allowDbsc)
        return { ok: true };
    try {
        const signBody = opts.signBody ?? false;
        let bodyBytes;
        if (signBody) {
            const ab = await req.clone().arrayBuffer();
            bodyBytes = new Uint8Array(ab);
        }
        await verifyBoundProof({
            sessionId: session.sessionId,
            proofHeader: req.headers.get("x-dbsc-bound-proof") ?? undefined,
            method: req.method,
            path: req.nextUrl.pathname,
            timestampWindowMs: opts.timestampWindowMs,
            signBody,
            bodyBytes,
            ...(opts.replayCache !== undefined && { replayCache: opts.replayCache }),
        }, opts.storage);
        return { ok: true };
    }
    catch (err) {
        if (err instanceof DbscVerificationError) {
            return {
                ok: false,
                response: NextResponse.json({ error: err.message, code: err.code }, { status: 403 }),
            };
        }
        throw err;
    }
}
