// @dbsc-toolkit/better-auth/dist/adapter.js
const CHALLENGE_PREFIX = "dbsc-challenge:";
function challengeIdentifier(jti) {
    return `${CHALLENGE_PREFIX}${jti}`;
}
export function createBetterAuthStorageAdapter(db, internalAdapter) {
    return {
        async getSession(id) {
            const row = await db.findOne({
                model: "dbscSession",
                where: [{ field: "id", value: id }],
            });
            return row ? rowToSession(row) : null;
        },
        async setSession(session) {
            const existing = await db.findOne({
                model: "dbscSession",
                where: [{ field: "id", value: session.id }],
            });
            if (existing) {
                await db.update({
                    model: "dbscSession",
                    where: [{ field: "id", value: session.id }],
                    update: sessionToRow(session),
                });
            }
            else {
                // Pass id with forceAllowId — Kysely adapter warns otherwise.
                // dbscSession.id is the Better Auth session id (canonical key).
                await db.create({
                    model: "dbscSession",
                    data: { id: session.id, ...sessionToRow(session) },
                    forceAllowId: true,
                });
            }
        },
        async deleteSession(id) {
            await db.delete({
                model: "dbscSession",
                where: [{ field: "id", value: id }],
            });
        },
        async getBoundKey(sessionId, kind) {
            if (kind) {
                const row = await db.findOne({
                    model: "dbscBoundKey",
                    where: [
                        { field: "sessionId", value: sessionId },
                        { field: "kind", value: kind },
                    ],
                });
                return row ? rowToBoundKey(row) : null;
            }
            const nativeRow = await db.findOne({
                model: "dbscBoundKey",
                where: [
                    { field: "sessionId", value: sessionId },
                    { field: "kind", value: "native" },
                ],
            });
            if (nativeRow)
                return rowToBoundKey(nativeRow);
            const boundRow = await db.findOne({
                model: "dbscBoundKey",
                where: [
                    { field: "sessionId", value: sessionId },
                    { field: "kind", value: "bound" },
                ],
            });
            return boundRow ? rowToBoundKey(boundRow) : null;
        },
        async setBoundKey(key) {
            const existing = await db.findOne({
                model: "dbscBoundKey",
                where: [
                    { field: "sessionId", value: key.sessionId },
                    { field: "kind", value: key.kind },
                ],
            });
            const data = boundKeyToRow(key);
            if (existing) {
                await db.update({
                    model: "dbscBoundKey",
                    where: [
                        { field: "sessionId", value: key.sessionId },
                        { field: "kind", value: key.kind },
                    ],
                    update: data,
                });
            }
            else {
                await db.create({ model: "dbscBoundKey", data });
            }
        },
        async deleteBoundKey(sessionId, kind) {
            if (kind) {
                await db.delete({
                    model: "dbscBoundKey",
                    where: [
                        { field: "sessionId", value: sessionId },
                        { field: "kind", value: kind },
                    ],
                });
            }
            else {
                await db.delete({
                    model: "dbscBoundKey",
                    where: [{ field: "sessionId", value: sessionId }],
                });
            }
        },
        async getChallenge(jti) {
            const row = await internalAdapter.findVerificationValue(challengeIdentifier(jti));
            if (!row)
                return null;
            return verificationRowToChallenge(jti, row);
        },
        async setChallenge(challenge) {
            await internalAdapter.createVerificationValue({
                identifier: challengeIdentifier(challenge.jti),
                value: JSON.stringify({ sessionId: challenge.sessionId, consumed: challenge.consumed }),
                expiresAt: new Date(challenge.expiresAt),
            });
        },
        async consumeChallenge(jti) {
            if (internalAdapter.consumeVerificationValue) {
                const result = await internalAdapter.consumeVerificationValue(challengeIdentifier(jti));
                return result !== null;
            }
            // Fallback: find + delete (less safe under concurrent requests, but functional)
            const row = await internalAdapter.findVerificationValue(challengeIdentifier(jti));
            if (!row)
                return false;
            if (internalAdapter.deleteVerificationValue) {
                await internalAdapter.deleteVerificationValue(row.id);
            }
            return true;
        },
        async revokeSession(sessionId) {
            await db.delete({
                model: "dbscBoundKey",
                where: [{ field: "sessionId", value: sessionId }],
            });
            await db.delete({
                model: "dbscSession",
                where: [{ field: "id", value: sessionId }],
            });
        },
        async revokeAllForUser(userId) {
            const sessions = await db.findMany({
                model: "dbscSession",
                where: [{ field: "userId", value: userId }],
            });
            for (const sess of sessions) {
                await db.delete({
                    model: "dbscBoundKey",
                    where: [{ field: "sessionId", value: sess["id"] }],
                });
            }
            await db.delete({
                model: "dbscSession",
                where: [{ field: "userId", value: userId }],
            });
        },
    };
}
function rowToSession(row) {
    return {
        id: String(row["id"]),
        userId: String(row["userId"]),
        tier: String(row["tier"]),
        createdAt: Number(row["createdAt"]),
        expiresAt: Number(row["expiresAt"]),
        lastRefreshAt: Number(row["lastRefreshAt"]),
    };
}
function sessionToRow(s) {
    return {
        userId: s.userId,
        tier: s.tier,
        createdAt: s.createdAt,
        expiresAt: s.expiresAt,
        lastRefreshAt: s.lastRefreshAt,
    };
}
function rowToBoundKey(row) {
    return {
        sessionId: String(row["sessionId"]),
        kind: String(row["kind"]),
        jwk: JSON.parse(String(row["jwk"])),
        createdAt: Number(row["createdAt"]),
        algorithm: String(row["algorithm"]),
    };
}
function boundKeyToRow(k) {
    return {
        sessionId: k.sessionId,
        kind: k.kind,
        jwk: JSON.stringify(k.jwk),
        createdAt: k.createdAt,
        algorithm: k.algorithm,
    };
}
function verificationRowToChallenge(jti, row) {
    const parsed = JSON.parse(row.value);
    return {
        jti,
        sessionId: parsed.sessionId,
        consumed: parsed.consumed,
        createdAt: 0,
        expiresAt: row.expiresAt.getTime(),
    };
}
