// @dbsc-toolkit/better-auth/dist/schema.js
/**
 * Better Auth schema extension for dbsc-toolkit.
 *
 * Defines two tables:
 *   dbscSession  — tracks binding state (tier, lastRefreshAt) per session
 *   dbscBoundKey — stores the JWK for native (TPM) and bound (polyfill) keys
 *
 * Challenges are stored in Better Auth's built-in `verification` table
 * so we get atomic consumeVerificationValue for free.
 */
export const dbscSchema = {
    dbscSession: {
        fields: {
            userId: {
                type: "string",
                required: true,
                references: { model: "user", field: "id" },
            },
            tier: {
                type: "string",
                required: true,
                // "dbsc" | "bound" | "none"
            },
            createdAt: {
                type: "number",
                required: true,
            },
            expiresAt: {
                type: "number",
                required: true,
            },
            lastRefreshAt: {
                type: "number",
                required: true,
            },
        },
    },
    dbscBoundKey: {
        fields: {
            sessionId: {
                type: "string",
                required: true,
                references: { model: "dbscSession", field: "id" },
            },
            kind: {
                type: "string",
                required: true,
                // "native" | "bound"
            },
            jwk: {
                type: "string",
                required: true,
                // JSON-serialized JWK — stored as text, never logged
            },
            createdAt: {
                type: "number",
                required: true,
            },
            algorithm: {
                type: "string",
                required: true,
                // "ES256" | "RS256"
            },
        },
    },
};
