// dbsc-toolkit/dist/storage/memory/index.js
export class MemoryStorage {
    sessions = new Map();
    // Keyed by `${sessionId}:${kind}`. A session can hold up to two rows
    // (one per kind) — see BoundKey.kind in core/types.ts.
    keys = new Map();
    challenges = new Map();
    revoked = new Set();
    async getSession(id) {
        const sess = this.sessions.get(id);
        if (!sess)
            return null;
        if (Date.now() > sess.expiresAt) {
            this.sessions.delete(id);
            return null;
        }
        return sess;
    }
    async setSession(session) {
        this.sessions.set(session.id, session);
    }
    async deleteSession(id) {
        this.sessions.delete(id);
        this.keys.delete(`${id}:native`);
        this.keys.delete(`${id}:bound`);
    }
    async getBoundKey(sessionId, kind) {
        if (kind)
            return this.keys.get(`${sessionId}:${kind}`) ?? null;
        return (this.keys.get(`${sessionId}:native`) ??
            this.keys.get(`${sessionId}:bound`) ??
            null);
    }
    async setBoundKey(key) {
        this.keys.set(`${key.sessionId}:${key.kind}`, key);
    }
    async deleteBoundKey(sessionId, kind) {
        if (kind) {
            this.keys.delete(`${sessionId}:${kind}`);
            return;
        }
        this.keys.delete(`${sessionId}:native`);
        this.keys.delete(`${sessionId}:bound`);
    }
    async getChallenge(jti) {
        const challenge = this.challenges.get(jti);
        if (!challenge)
            return null;
        if (Date.now() > challenge.expiresAt) {
            this.challenges.delete(jti);
            return null;
        }
        return challenge;
    }
    async setChallenge(challenge) {
        this.challenges.set(challenge.jti, challenge);
    }
    async consumeChallenge(jti) {
        const challenge = this.challenges.get(jti);
        if (!challenge || challenge.consumed)
            return false;
        challenge.consumed = true;
        return true;
    }
    async revokeSession(sessionId) {
        this.revoked.add(sessionId);
        this.sessions.delete(sessionId);
        this.keys.delete(`${sessionId}:native`);
        this.keys.delete(`${sessionId}:bound`);
    }
    async revokeAllForUser(userId) {
        for (const [id, sess] of this.sessions.entries()) {
            if (sess.userId === userId) {
                this.revoked.add(id);
                this.sessions.delete(id);
                this.keys.delete(`${id}:native`);
                this.keys.delete(`${id}:bound`);
            }
        }
    }
    // Useful in tests: removes all expired entries from all maps
    gc() {
        const now = Date.now();
        for (const [id, sess] of this.sessions.entries()) {
            if (now > sess.expiresAt)
                this.sessions.delete(id);
        }
        for (const [jti, ch] of this.challenges.entries()) {
            if (now > ch.expiresAt)
                this.challenges.delete(jti);
        }
    }
}
export { MemoryReplayCache } from "./replay-cache.js";
