// dbsc-toolkit/dist/koa/require-proof.js
import { verifyBoundProof, noBindingReason, DbscVerificationError, } from "../core/index.js";
import { DBSC_INTERNAL } from "../node/index.js";
async function readRawBody(ctx) {
    const rawBody = ctx.request.rawBody;
    if (typeof rawBody === "string")
        return new TextEncoder().encode(rawBody);
    const req = ctx.req;
    const cached = req["__dbscRawBody"];
    if (Buffer.isBuffer(cached))
        return new Uint8Array(cached);
    const chunks = [];
    for await (const chunk of req)
        chunks.push(chunk);
    const raw = Buffer.concat(chunks);
    req["__dbscRawBody"] = raw;
    return new Uint8Array(raw);
}
/**
 * The route guard for Koa. Mount `dbsc()` first so the session is on
 * `ctx.state.dbsc`. Reads the raw body from `ctx.request.rawBody` when a body
 * parser populated it, else from the socket.
 */
export function requireProof(opts = {}) {
    return async (ctx, next) => {
        const session = ctx.state.dbsc;
        const tier = session?.tier ?? "none";
        const skipped = session?.skipped ?? [];
        if (!session?.sessionId || tier === "none") {
            ctx.status = 403;
            ctx.body = { error: "device-bound session required", currentTier: "none", reason: noBindingReason(skipped), skipped };
            return;
        }
        const internal = session[DBSC_INTERNAL];
        const storage = opts.storage ?? internal?.storage;
        if (!storage) {
            ctx.status = 500;
            ctx.body = { error: "requireProof: storage unavailable — mount dbsc() before this route, or pass { storage }" };
            return;
        }
        const allowDbscWithoutProof = opts.allowDbscWithoutProof ?? (internal?.boundEnabled === false ? true : undefined);
        if (tier === "dbsc" && allowDbscWithoutProof) {
            await next();
            return;
        }
        const proofHeader = ctx.get("x-dbsc-bound-proof") || undefined;
        try {
            await verifyBoundProof({
                sessionId: session.sessionId,
                proofHeader,
                method: ctx.method,
                path: ctx.path,
                signBody: true,
                bodyBytes: await readRawBody(ctx),
                ...(opts.timestampWindowMs !== undefined && { timestampWindowMs: opts.timestampWindowMs }),
                ...(internal?.replayCache !== undefined && { replayCache: internal.replayCache }),
            }, storage);
            await next();
        }
        catch (err) {
            if (err instanceof DbscVerificationError) {
                ctx.status = 403;
                ctx.body = { error: err.message, code: err.code };
                return;
            }
            throw err;
        }
    };
}
